core/src/main/java/com/linecorp/armeria/client/routing/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -0,0 +1,163 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.routing.endpoint.healthcheck;
+
+import static com.linecorp.armeria.common.util.Functions.voidFunction;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import com.google.common.collect.ImmutableList;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.routing.EndpointGroup;
+
+import jp.skypencil.guava.stream.GuavaCollectors;
+
+/**
+ * An {@link EndpointGroup} decorator that only provides healthy {@link Endpoint}s.
+ */
+public abstract class HealthCheckedEndpointGroup implements EndpointGroup {
+    public static final long DEFAULT_HEALTHCHECK_RETRY_DELAY_MILLS = 3_000; // 3 seconds.
+
+    private final ClientFactory clientFactory;
+    private final EndpointGroup delegate;
+    private final long healthCheckRetryDelayMills;
+
+    volatile List<ServerConnection> allServers = ImmutableList.of();
+    volatile List<ServerConnection> healthyServers = ImmutableList.of();
+
+    /**
+     * Creates a new instance.
+     * A subclass being initialized with this constructor must call {@link #init()} before start being used.
+     */
+    protected HealthCheckedEndpointGroup(ClientFactory clientFactory,
+                                         EndpointGroup delegate,
+                                         long healthCheckRetryDelayMills) {
+        this.clientFactory = requireNonNull(clientFactory, "clientFactory");
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.healthCheckRetryDelayMills = healthCheckRetryDelayMills;
+    }
+
+    /**
+     * Update healty servers and start to schedule healthcheck.
+     * A subclass being initialized with this constructor must call {@link #init()} before start being used.
+     */
+    protected void init() {
+        checkAndUpdateHealthyServers();
+        scheduleCheckAndUpdateHealthyServers();
+    }
+
+    protected ClientFactory clientFactory() {
+        return clientFactory;
+    }
+
+    private void scheduleCheckAndUpdateHealthyServers() {
+        clientFactory.eventLoopGroup().schedule(() -> {
+            checkAndUpdateHealthyServers().thenRun(this::scheduleCheckAndUpdateHealthyServers);
+        }, healthCheckRetryDelayMills, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Update the servers this health checker client talks to.
+     */
+    private List<ServerConnection> updateServerList() {
```

Comment:
How about moving this method below its caller (`checkAndUpdateHealthyServer()`) for readability?